### PR TITLE
ZK-4330: provide select-all checkbox in tree component

### PR DESCRIPTION
**Pull request by b952426**
ZK-4330: provide select-all checkbox in tree component

### CODE REVIEW

File: zktest/src/archive/test2/F96-ZK-4330.zul [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            <treecols>
+                <treecol label="tree1"/>
+            </treecols>
+        </tree>
+        <tree id="tree2" multiple="true" checkmark="true" height="300px" width="200px">
+            <treecols>
+                <treecol label="tree2"/>
+            </treecols>
+        </tree>
+        <custom-attributes org.zkoss.zul.tree.initRodSize="-1"/>
```
Comment (b952426) [inline zktest/src/archive/test2/F96-ZK-4330.zul:69]: put into tree3

File: zul/src/archive/web/js/zul/sel/Treecol.js [hunk 2]
```diff
@@ -143,9 +154,51 @@ zul.sel.Treecol = zk.$extends(zul.mesh.SortWidget, {
...
+		var tree = this.getTree();
+		return tree != null && this.parent.firstChild == this
+			&& tree._checkmark && tree._multiple && !tree._tree$noSelectAll;
+	},
+	//@Override
 	domLabel_: function () {
 		return zUtl.encodeXML(this.getLabel(), {maxlength: this._maxlength});
-	}
+	},
+	_doClick: function (evt) {
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/Treecol.js:175]: doCmClick

File: zul/src/archive/web/js/zul/sel/Treecol.js [hunk 3]
```diff
@@ -143,9 +154,51 @@ zul.sel.Treecol = zk.$extends(zul.mesh.SortWidget, {
...
+			$n = jq(cm);
+		if (this._checked) {
+			$n.addClass(this.$s('checked'));
+			tree.selectAll(true, evt);
+		} else {
+			$n.removeClass(this.$s('checked'));
+			tree._select(null, evt);
+		}
+		tree.fire('onCheckSelectAll', this._checked, {toServer: true});
+
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/Treecol.js:188]: delete
remove toServer? use important

File: zul/src/org/zkoss/zul/Tree.java [hunk 4]
```diff
@@ -2852,6 +2858,19 @@ public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
 			final Map<String, Object> data = request.getData();
 			_anchorTop = AuRequests.getInt(data, "top", 0);
 			_anchorLeft = AuRequests.getInt(data, "left", 0);
+		} else if (cmd.equals("onCheckSelectAll")) {
+			CheckEvent evt = CheckEvent.getCheckEvent(request);
```
Comment (b952426) [inline zul/src/org/zkoss/zul/Tree.java:2862]: move to if

File: zul/src/org/zkoss/zul/Tree.java [hunk 5]
```diff
@@ -2852,6 +2858,19 @@ public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
...
+			if (_model instanceof Selectable) {
+				final Selectable<Object> smodel = (Selectable) _model;
+				SelectionControl control = smodel.getSelectionControl();
+				if (control != null) {
+					control.setSelectAll(evt.isChecked());
+				}
+			}
+		} else if (_model != null && cmd.equals("onCheckSelectRow")) {
```
Comment (b952426) [inline zul/src/org/zkoss/zul/Tree.java:2870]: _model instanceof Selectable

File: zul/src/archive/web/js/zul/sel/Treerow.js [hunk 6]
```diff
@@ -69,13 +69,16 @@ zul.sel.Treerow = zk.$extends(zul.Widget, {
...
+			tg = evt.domTarget,
+			tree = this.getTree();
 		if (tg == this.$n('open') || tg == this.$n('icon')) {
 			ti.setOpen(!ti._open);
 			evt.stop();
-			this.getTree().focus();
+			tree.focus();
 		} else if (!ti.isDisabled())
 			this.$supers('doClick_', arguments);
+			tree.fire('onCheckSelectRow', this.parent.isSelected(), {toServer: true});
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/Treerow.js:80]: onSelect

File: zul/src/archive/web/js/zul/sel/SelectWidget.js [hunk 7]
```diff
@@ -1140,10 +1140,14 @@ zul.sel.SelectWidget = zk.$extends(zul.mesh.MeshWidget, {
 			if (this.$$selectAll != undefined) {
 				return this.$$selectAll;
 			}
+			if (this.isAllRowSelected != undefined) {
+				return this.isAllRowSelected;
+			}
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/SelectWidget.js:1145]: select widget: onUpdateSelectAll

File: zul/src/archive/web/js/zul/sel/SelectWidget.js [hunk 8]
```diff
@@ -1140,10 +1140,14 @@ zul.sel.SelectWidget = zk.$extends(zul.mesh.MeshWidget, {
...
 		}
 
-		var isGroupSelect = this.groupSelect;
-		for (var it = this.getBodyWidgetIterator({skipHidden: true}), w; (w = it.next());) {
+		var isGroupSelect = this.groupSelect,
+			skipHidden = this.$instanceof(zul.sel.Tree) && !this._model ? false : true;
+		for (var it = this.getBodyWidgetIterator({skipHidden: skipHidden}), w; (w = it.next());) {
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/SelectWidget.js:1150]: other issue

File: zul/src/org/zkoss/zul/Select.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.zkoss.json.JSONAware;
+import org.zkoss.zk.au.out.AuInvoke;
+import org.zkoss.zk.ui.util.Clients;
+import org.zkoss.zul.ext.Selectable;
+import org.zkoss.zul.ext.SelectionControl;
+import org.zkoss.zul.impl.MeshElement;
+
+
+public abstract class Select extends MeshElement {
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Select.java:22]: at this moment just open an issue

File: zul/src/archive/web/js/zul/sel/Treeitem.js [hunk 10]
```diff
@@ -177,6 +177,7 @@ zul.sel.Treeitem = zk.$extends(zul.sel.ItemWidget, {
 			this._showKids(open);
 			if (open)
 				zUtl.fireShown(this);
+				tree._updHeaderCM();
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/sel/Treeitem.js:180]: if (open) {

}

**Review by scribetw — APPROVED** [review]
Open an issue about SelectElement

File: zul/src/org/zkoss/zul/Tree.java [hunk 11]
```diff
@@ -2852,6 +2870,26 @@ public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
...
 			_anchorLeft = AuRequests.getInt(data, "left", 0);
+		} else if (cmd.equals("onCheckSelectAll") && _model != null && _model instanceof Selectable) {
+			CheckEvent evt = CheckEvent.getCheckEvent(request);
+			final Selectable<Object> selectableModel = (Selectable<Object>) _model;
+			SelectionControl control = selectableModel.getSelectionControl();
+			if (control == null)
+				throw new IllegalStateException(
+						"SelectionControl cannot be null, please implement SelectionControl interface for SelectablModel");
+			control.setSelectAll(evt.isChecked());
+		} else if (cmd.equals("onUpdateSelectAll") && _model != null && _model instanceof Selectable) {
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Tree.java:2881]: var b = _model instanceof Selectable

**Review by scribetw — CHANGES_REQUESTED** [review]
extract condition

File: zul/src/archive/web/js/zul/sel/Treecol.js [hunk 12]
```diff
@@ -143,9 +154,49 @@ zul.sel.Treecol = zk.$extends(zul.mesh.SortWidget, {
...
+		var tree = this.getTree(),
+			cm = this.$n('cm');
+		if (tree && tree._checkmark) {
+			var n = evt.domTarget;
+			if (n == cm || n.parentNode == cm) {
+				return;
+			}
+			this.$supers('doClick_', arguments);
+		}
+	},
```
Comment (b952426) [inline zul/src/archive/web/js/zul/sel/Treecol.js:199]: }

**Review by plijyfes — APPROVED** [review]

**Review by scribetw — APPROVED** [review]

**Pull request merged**